### PR TITLE
(WIP) feat: Implement Occurrences data transformation with `dbt`

### PR DESCRIPTION
**Pull request by kritchie**
# Context

In the context of the radiant project, we need to be able to import data from the data lake into StarRocks to enable the downstream apps to perform fast lookup (and queries) on the data.

The data lake contains several tables which are the results of bioinformatics processes. 

The current PR focuses defining the necessary transformations to deliver the `occurences`-related tables. 

# Contribution

This contribution adds the following models: 

- `stg_locus_lookup`: Contains the locus key/mapping to speed up locus join operations. 
- `stg_normalized_snv`: Contains the transformed data from 

:information_source: Models' documentation and specific testing for each field will be added in a follow-up PR

This required the following design choices: 

- Implemented a workaround using `uuid_numeric` to generate `LARGEINT`-based unique ids instead of `BIGINT AUTO_INCREMENT` which is not supported in `dbt`. 
- Changed the original `sha2(...)` hashing function to use a faster `md5` function. (Inspired from dbt's surrogate keys).
- Added a specific `is_unique` data test to workaround the failing default `unique` dbt test. This seems to be a corner case of using `uuid_numeric` to generate largeints (hypothesis at this point, will need to be thoroughly investigated).  

### Generated Docs

image

### Running logs

The run with the current filters takes around ~20 minutes.

```
16:31:57  Running with dbt=1.8.9
16:31:57  Registered adapter: starrocks=1.9.0
16:31:58  Found 2 models, 7 data tests, 2 sources, 426 macros
16:31:58  
16:31:58  Concurrency: 1 threads (target='dev')
16:31:58  
16:31:58  1 of 9 START sql table model test.stg_locus_lookup ............................. [RUN]
16:31:59  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 2 seconds...
16:32:01  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 4 seconds...
16:32:10  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 8 seconds...
16:32:18  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 16 seconds...
16:32:34  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 32 seconds...
16:33:09  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 64 seconds...
16:34:13  starrocks adapter: Task 891b3b7dbb6f47d2ad8c8db0d5a6f790 progress [0%]. Waiting 128 seconds...
16:36:21  starrocks adapter: Task [891b3b7dbb6f47d2ad8c8db0d5a6f790] finished with status [SUCCESS]
16:36:21  1 of 9 OK created sql table model test.stg_locus_lookup ........................ [SUCCESS 1 in 262.80s]
16:36:21  2 of 9 START test is_unique_stg_locus_lookup_locus ............................. [RUN]
16:36:23  2 of 9 PASS is_unique_stg_locus_lookup_locus ................................... [PASS in 1.89s]
16:36:23  3 of 9 START test is_unique_stg_locus_lookup_locus_id .......................... [RUN]
16:36:25  3 of 9 PASS is_unique_stg_locus_lookup_locus_id ................................ [PASS in 1.33s]
16:36:25  4 of 9 START test not_null_stg_locus_lookup_locus .............................. [RUN]
16:36:25  4 of 9 PASS not_null_stg_locus_lookup_locus .................................... [PASS in 0.22s]
16:36:25  5 of 9 START test not_null_stg_locus_lookup_locus_id ........................... [RUN]
16:36:25  5 of 9 PASS not_null_stg_locus_lookup_locus_id ................................. [PASS in 0.55s]
16:36:25  6 of 9 START sql table model test.stg_normalized_snv ........................... [RUN]
16:36:26  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [0%]. Waiting 2 seconds...
16:36:28  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 4 seconds...
16:36:32  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 8 seconds...
16:36:41  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 16 seconds...
16:36:57  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 32 seconds...
16:37:29  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 64 seconds...
16:38:33  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 128 seconds...
16:40:42  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 256 seconds...
16:44:58  starrocks adapter: Task 3f43da91012047ca898cc7f7e475e5e1 progress [3%]. Waiting 512 seconds...
16:53:30  starrocks adapter: Task [3f43da91012047ca898cc7f7e475e5e1] finished with status [SUCCESS]
16:53:31  6 of 9 OK created sql table model test.stg_normalized_snv ...................... [SUCCESS 1 in 1025.40s]
16:53:31  7 of 9 START test not_null_stg_normalized_snv_locus_id ......................... [RUN]
16:53:31  7 of 9 PASS not_null_stg_normalized_snv_locus_id ............................... [PASS in 0.50s]
16:53:31  8 of 9 START test not_null_stg_normalized_snv_part ............................. [RUN]
16:53:31  8 of 9 PASS not_null_stg_normalized_snv_part ................................... [PASS in 0.21s]
16:53:31  9 of 9 START test not_null_stg_normalized_snv_seq_id ........................... [RUN]
16:53:34  9 of 9 PASS not_null_stg_normalized_snv_seq_id ................................. [PASS in 2.22s]
16:53:34  
16:53:34  Finished running 2 table models, 7 data tests in 0 hours 21 minutes and 36.24 seconds (1296.24s).
16:53:34  
16:53:34  Completed successfully
16:53:34  
16:53:34  Done. PASS=9 WARN=0 ERROR=0 SKIP=0 TOTAL=9
```

## :grey_question: Open Questions
- Do we really want to hash the locus ? If yes, which function should we use ? 
- Should we implement the `incremental` materialisation at this stage ?

### CODE REVIEW

File: radiant/sjra-62/dbt/models/staging/stg_normalized_snv.sql [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        o.ad_alt,
+        o.dp,
+        o.gq,
+
+        -- Variant characteristics
+        o.chromosome,
+        o.start,
+        o.zygosity,
+        o.has_alt,
+        o.quality,
```
Comment (jecos) [inline radiant/sjra-62/dbt/models/staging/stg_normalized_snv.sql]: Quality is not necessary

File: radiant/sjra-62/dbt/models/staging/stg_normalized_snv.sql [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+),
+
+-- Final combined dataset with variant information
+final as (
+    select
+        s.part as part,
+        s.seq_id as seq_id,
+
+        dict_mapping('variant_dict',
+            sha2(concat_ws('-', chromosome, start, reference, alternate), 256)
```
Comment (jecos) [inline radiant/sjra-62/dbt/models/staging/stg_normalized_snv.sql]: open discussion : should we use a sha as a mapping key or directly the locus (risk of collusion?)

Comment (kritchie) [reply radiant/sjra-62/dbt/models/staging/stg_normalized_snv.sql]: Good question, it might be a good idea since we good potentially re-use those concatenated strings directly when displaying information to the end-users. They could see the actual values that make up the locus as opposed to a hash which doesn't provide much information. 

That being said, there might some extra considerations like the string length that we allow that could make this unusable. 

Otherwise, the `dbt-utils` package provides a [`generate_surrogate_key`](https://github.com/dbt-labs/dbt-utils/blob/main/macros/sql/generate_surrogate_key.sql) macro that essentially concatenates the fields together and apply a `md5`, which is faster than a `sha2` to compute, but has a smaller range of values (still good enough for the use case we have in my opinion). 

So tldr; Good question, let's have a chat offline about this :+1:

File: radiant/sjra-62/dbt/tests/generic/is_unique.sql
Comment (kritchie) [inline radiant/sjra-62/dbt/tests/generic/is_unique.sql:1]: For some reason, the `unique` default test fails with those tables with a StarRocks error: 

```
 Database Error in test unique_stg_locus_lookup_locus (models/staging/schema.yml)
  1064 (HY000): Constant 1 can't compare with Constant 1
  compiled code at target/run/sjra62/models/staging/schema.yml/unique_stg_locus_lookup_locus.sql
``` 

This simple macro was added to replace the default `unique` test, but this will need to be investigated.

File: radiant/sjra-62/dbt/macros/generate_surrogate_key.sql
Comment (kritchie) [inline radiant/sjra-62/dbt/macros/generate_surrogate_key.sql:1]: This macro was adapted from https://github.com/dbt-labs/dbt-utils/blob/main/macros/sql/generate_surrogate_key.sql

I felt it was not necessary to include `dbt-utils` as a dbt dependencies just to use the surrogate key feature. 

If we decide to stick with the `sha2` we can remove this macro.

File: radiant/sjra-62/dbt/models/staging/stg_locus_lookup.sql [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+with locuses as (
+
+    select distinct {{ generate_surrogate_key(['chromosome', 'start', 'reference', 'alternate']) }} as locus
+    from {{ source('cqdg_datalake', 'normalized_snv') }}
+),
+
+combined as (
+    select
+        l.locus as locus,
+        uuid_numeric() as locus_id
```
Comment (kritchie) [inline radiant/sjra-62/dbt/models/staging/stg_locus_lookup.sql]: This is different from the `BIGINT AUTO_INCREMENT` implementation. 
The [`uuid_numeric`](https://docs.starrocks.io/docs/sql-reference/sql-functions/utility-functions/uuid_numeric) creates a `LARGEINT` (16 bytes). 

This enables the creation of `stg_locus_lookup` through `dbt`.

(Generated the whole bunch of `uuid` in about ~5 minutes)

File: radiant/sjra-62/dbt/tests/generic/is_unique.sql
Comment (kritchie) [inline radiant/sjra-62/dbt/tests/generic/is_unique.sql:1]: For some reason, the unique default test fails with those tables with a StarRocks error:

 Database Error in test unique_stg_locus_lookup_locus (models/staging/schema.yml)
  1064 (HY000): Constant 1 can't compare with Constant 1
  compiled code at target/run/sjra62/models/staging/schema.yml/unique_stg_locus_lookup_locus.sql
This simple macro was added to replace the default unique test, but this will need to be investigated.

File: radiant/sjra-62/dbt/dbt_project.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    - "target"
+    - "dbt_modules"
+    - "logs"
+
+require-dbt-version: [">=1.0.0", "<2.0.0"]
+
+models:
+  sjra62:
+    staging:
+      materialized: table
```
Comment (kritchie) [inline radiant/sjra-62/dbt/dbt_project.yml:35]: This should change eventually to supported incremental updates with dynamic overwrites to this table.

File: radiant/sjra-62/dbt/tests/generic/is_unique.sql
Comment (kritchie) [inline radiant/sjra-62/dbt/tests/generic/is_unique.sql:1]: For some reason, the unique default test fails with those tables with a StarRocks error:

```
 Database Error in test unique_stg_locus_lookup_locus (models/staging/schema.yml)
  1064 (HY000): Constant 1 can't compare with Constant 1
  compiled code at target/run/sjra62/models/staging/schema.yml/unique_stg_locus_lookup_locus.sql
```

This simple macro was added to replace the default unique test, but this will need to be investigated.
